src/engine/DebugOverlays.cs

```diff
@@ -0,0 +1,164 @@
+﻿using Godot;
+
+/// <summary>
+///   Main script for debugging.
+///   Partial class: Override functions, debug panel
+/// </summary>
+public partial class DebugOverlays : Control
+{
+    [Export]
+    public NodePath FPSCheckBoxPath = null!;
+
+    [Export]
+    public NodePath PerformanceMetricsCheckBoxPath = null!;
+
+    [Export]
+    public NodePath DebugPanelDialogPath = null!;
+
+    [Export]
+    public NodePath FPSCounterPath = null!;
+
+    [Export]
+    public NodePath PerformanceMetricsPath = null!;
+
+    [Export]
+    public NodePath EntityLabelsPath = null!;
+
+    private static DebugOverlays? instance;
+
+    private CustomDialog debugPanelDialog = null!;
+    private CustomCheckBox fpsCheckBox = null!;
+    private CustomCheckBox performanceMetricsCheckBox = null!;
+    private Control fpsCounter = null!;
+    private CustomDialog performanceMetrics = null!;
+    private Control labelsLayer = null!;
+
+    private DebugOverlays()
+    {
+        instance = this;
+    }
+
+    public static DebugOverlays Instance => instance ?? throw new InstanceNotLoadedYetException();
+
+    public override void _EnterTree()
+    {
+        base._EnterTree();
+
+        Show();
+        InputManager.RegisterReceiver(this);
+
+        // Entity label
+        var rootTree = GetTree();
+        rootTree.Connect("node_added", this, nameof(OnNodeAdded));
+        rootTree.Connect("node_removed", this, nameof(OnNodeRemoved));
```

Comment:
This should really only be connected while debugging is active. I do not want the debugging features to use *any* performance when not in use.

Reply:
But this means we need to know all the entities when entity label is turned on, which requires much time.

Reply:
In my opinion that's better than requiring all the tens of thousands of players to have a slower experience who never turn on the debug overlay.